rskj-core/src/main/java/org/ethereum/rpc/LogFilter.java

```diff
@@ -68,40 +67,44 @@ void onLogMatch(LogInfo logInfo, Block b, int txIndex, Transaction tx, int logId
         add(new LogFilterEvent(new LogFilterElement(logInfo, b, txIndex, tx, logIdx)));
     }
 
-    void onTransaction(Transaction tx, Block b, int txIndex) {
-        TransactionInfo txInfo = blockchain.getTransactionInfo(tx.getHash().getBytes());
+    void onTransaction(Transaction tx, Block block, int txIndex, boolean reverseLogIdxOrder) {
+        TransactionInfo txInfo = blockchain.getTransactionInfoByBlock(tx, block.getHash().getBytes());
         TransactionReceipt receipt = txInfo.getReceipt();
 
         LogFilterElement[] logs = new LogFilterElement[receipt.getLogInfoList().size()];
 
         for (int i = 0; i < logs.length; i++) {
-            LogInfo logInfo = receipt.getLogInfoList().get(i);
+            int logIdx = reverseLogIdxOrder ? logs.length - i - 1 : i;
+
+            LogInfo logInfo = receipt.getLogInfoList().get(logIdx);
 
             if (addressesTopicsFilter.matchesExactly(logInfo)) {
-                onLogMatch(logInfo, b, txIndex, receipt.getTransaction(), i);
+                onLogMatch(logInfo, block, txIndex, receipt.getTransaction(), logIdx);
             }
         }
     }
 
-    void onBlock(Block b) {
-        if (addressesTopicsFilter.matchBloom(new Bloom(b.getLogBloom()))) {
-            int txIdx = 0;
+    void onBlock(Block block, boolean reverseTxOrder) {
+        if (!addressesTopicsFilter.matchBloom(new Bloom(block.getLogBloom()))) {
+            return;
+        }
 
-            for (Transaction tx : b.getTransactionsList()) {
-                onTransaction(tx, b, txIdx);
-                txIdx++;
-            }
+        List<Transaction> txs = block.getTransactionsList();
+
+        for (int i = 0; i < txs.size(); i++) {
+            int txIdx = reverseTxOrder ? txs.size() - i - 1 : i;
+            onTransaction(txs.get(txIdx), block, txIdx, reverseTxOrder);
         }
     }
 
     @Override
     public void newBlockReceived(Block b) {
         if (this.fromLatestBlock) {
             this.clearEvents();
-            onBlock(b);
+            onBlock(b, false);

```

Comment:
isn't reverseTxOrder (true) should be specified here?

Reply:
I might be missing sth because the logic is a bit complex, but I think it's correct for it to be `false`, as when we receive a new block we always add it forward to the bloom.

Bear in mind that `org.ethereum.rpc.LogFilter#retrieveHistoricalData` is static and receives its own (and newly created) `LogFilter`. This filter is then added to `installedFilters` list on `org.ethereum.rpc.FilterManager#registerFilter`. This list is then looped and ends up calling `org.ethereum.rpc.FilterManager#newBlockReceived` that itself calls `org.ethereum.rpc.LogFilter#newBlockReceived` for each installed filter.

Furthermore, we are only adding items backwards if the bloom range does not exist already for the given filter.